File: python/src/threads/agents/usd.py
"""`usd()` (spec/api.json): dollars as the integer nano-dollars budgets and prices use."""

import math

from threads.agents.config import ConfigError

_MAX_SAFE = 2**53 - 1


def usd(dollars: float) -> int:
    """usd(0.5) == 500_000_000, rounded half up to the nearest nano-dollar (the same IEEE
    operations as TypeScript's). Raises ConfigError invalid_config for a negative, non-finite
    or out-of-range amount."""
    if isinstance(dollars, bool) or not math.isfinite(dollars) or dollars < 0:
        raise ConfigError("invalid_config", f"usd needs a finite amount >= 0, got {dollars!r}")
    nanos = math.floor(dollars * 1_000_000_000 + 0.5)
    if nanos > _MAX_SAFE:
        raise ConfigError("invalid_config", f"usd({dollars!r}) is more nano-dollars than 2**53-1")
    return nanos

File: typescript/packages/core/src/agent/usd.ts
import { ConfigError } from "./errors";

/**
 * spec/api.json usd: a US-dollar amount as the integer nano-dollars budgets and prices use,
 * rounded half up (usd(0.5) is 500_000_000). Throws ConfigError for a negative, non-finite or
 * too large amount.
 */
export function usd(dollars: number): number {
  const nanos = Math.floor(dollars * 1e9 + 0.5);
  if (!Number.isFinite(dollars) || dollars < 0 || !Number.isSafeInteger(nanos))
    throw new ConfigError(
      "invalid_config",
      `usd(${dollars}): give a finite, non-negative amount of at most ${Number.MAX_SAFE_INTEGER / 1e9} dollars`,
    );
  return nanos;
}
